lib/reservoir-sampler.js

```diff
@@ -5,56 +5,68 @@ var es = require('event-stream');
 var _defaults = require('lodash.defaults');
 var debug = require('debug')('mongodb-collection-sample:reservoir-sampler');
 
+var RESERVOIR_SAMPLE_LIMIT = 10000;
 
 /**
- * Take an `_id` and emit the source document.
+ * Take `_id` values, buffer them into a batch, run $in query on a batch
+ * and emit the source documents.
  *
  * @param {mongodb.DB} db
  * @param {String} collectionName to source from.
  * @param {Object} opts
  * @option {Object} fields to return for each document [default: `null`].
  * @return {stream.Transform}
+ *
  * @api private
  */
-function _idToDocument(db, collectionName, opts) {
+function batchDocs(db, collectionName, size, opts) {
   opts = _defaults(opts || {}, {
     fields: null,
+    batchSize: Math.sqrt(size),
```

Comment:
Forgot that the hard limit here was ten thousand, not one thousand... :) Going up to 10K took an additional 54ms. Maybe at some point we will need to batch these? However if we do, I don't think square root is the best way as any size less than 1000 is wasting time sending multiple requests over the network.

If we do batch them, I'd prefer going with fixed size buckets of 2,000 or so and calling it good. For me 2k _ids only takes 14ms; still quite fast.